---
title: "Metabolomics"
author: "Johannes Rainer<br><strong>Eurac Research</strong>, Bolzano, Italy<br>johannes.rainer@eurac.edu - github: jorainer - twitter: jo_rainer"
date: "CSAMA 2019"
output: 
  ioslides_presentation:
    widescreen: false
    fig_width: 7
    fig_height: 5
    fig_retina: 2
    fig_caption: false
    transition: faster
    css: jostyle.css
---

<style type="text/css">

slides > slide:not(.nobackground):after {
  content: '';
}

slides > slide {
    -webkit-transition:none !important;transition:none !important;
}

.build > * {
  -webkit-transition: opacity 0.1s ease-in-out;
  -webkit-transition-delay: 0.1s;
  -moz-transition: opacity 0.1s ease-in-out 0.1s;
  -o-transition: opacity 0.1s ease-in-out 0.1s;
  transition: opacity 0.1s ease-in-out 0.1s;
}

</style>

<!--
Export to pdf:
webshot function from webshot package.
Might have to change stuff in the style sheet to make it working.
library(webshot)
install_phantomjs()
fn <- paste0("file://", normalizePath("metabolomics.html"))
rmdshot("metabolomics.Rmd", "metabolomics.pdf")
-->

## Content

- (Brief) introduction to metabolomics
- Preprocessing of LC-MS data
- Normalization
- Annotation/identification

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_01.png")
```

<div style="font-size: 80%;">
- Key metabolic pathway common to all cells.
- Creates energy by converting glucose to pyruvate.
<div>

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_02.png")
```

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_04.png")
```

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_06.png")
```

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_10.png")
```

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_11.png")
```

## Metabolite? Metabolism?

- [Glycolysis](https://en.wikipedia.org/wiki/Glycolysis)

```{r out.width = "750px", echo = FALSE}
knitr::include_graphics("images/glycolysis_12.png")
```

- Metabolites: intermediates and products of cellular processes.

## Metabolomics? {.build}

- Large-scale study of small molecules (metabolites) in a system (cell, tissue,
  organism).

<div>
- Comparison of the different -omes:
</div>
<div>
  - **Genome**: what can happen.
</div>
<div>
  - **Transcriptome**: what appears to be happening.
</div>
<div>
  - **Proteome**: what makes it happen.
</div>
<div>
  - **Metabolome**: what actually happened.
</div>
<div>
- Metabolome influenced by genetic **and** environmental factors.
</div>

## How can we measure metabolites? {.build}

<div>
- Nuclear Magnetic Resonance (NMR) - not covered here.
- Mass spectrometry (MS)-based metabolomics.
</div>

<div>
- Metabolites small enough to be directly measured by MS.
- Most metabolites uncharged - need to create ions first.
</div>

<!-- <div> -->
<!-- - Targeted/untargeted metabolomics: -->
<!-- </div> -->
<!-- <div> -->
<!--   - **targeted**: quantitative measurement of selected metabolites. -->
<!-- </div> -->
<!-- <div> -->
<!--   - **untargeted**: semi-quantitative measurement of all metabolites (detectable -->
<!--     with the setup) in a sample. -->
<!-- </div> -->


## Mass Spectrometry (MS)

<div style="position:absolute; top:28%; left:7%; width:400px">
```{r out.width = "400px", echo = FALSE}
knitr::include_graphics("images/MS_ionization_mod.png")
```
</div>

## Mass Spectrometry (MS)

<div style="position:absolute; top:28%; left:7%; width:400px">
```{r out.width = "400px", echo = FALSE}
knitr::include_graphics("images/MS_ionization_mod.png")
```
</div>
<div style="position:absolute; top:20%; left:53%; width:300px">
```{r out.width = "300px", echo = FALSE}
knitr::include_graphics("images/Spectrum1_mod.png")
```
</div>

<div style="position:absolute; top:65%;left:7%">
> - **Problem:** unable to distinguish between metabolites with the same/similar 
    mass-to-charge ratio (m/z).
> - **Solution:** additional separation of metabolites prior to MS.
</div>


## Liquid chromatography {.build}

<div>
- Sample is dissolved in a fluid (mobile phase).

<div style="position:absolute; top:75%;left:25%">
```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/LC_01.png")
```
</div>
</div>

## Liquid chromatography

- Sample is dissolved in a fluid (mobile phase).

- Mobile phase carries analytes through column (stationary phase).

<div style="position:absolute; top:75%;left:25%">
```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/LC_02.png")
```
</div>


## Liquid chromatography

- Sample is dissolved in a fluid (mobile phase).

- Mobile phase carries analytes through column (stationary phase).

- Separation based on affinity for the column's stationary phase.

<div style="position:absolute; top:75%;left:25%">
```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/LC_03.png")
```
</div>


## Liquid chromatography

- Sample is dissolved in a fluid (mobile phase).

- Mobile phase carries analytes through column (stationary phase).

- Separation based on affinity for the column's stationary phase.

- [HILIC](https://en.wikipedia.org/wiki/Hydrophilic_interaction_chromatography)
(hyrophilic liquid interaction chromatography):

  - Hydrophilic, polar stationary phase.
  
  - Analytes solved in mobile phase.
  
  - Analytes separated by polarity: compounds with low polarity elute first,
    with high polarity later.

<div style="position:absolute; top:75%;left:25%">
```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/LC_03.png")
```
</div>


<!-- Now comes the incremental LC-MS description -->
## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/MS_Spectrum1.png")
```

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum1.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum2.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum3.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum4.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum5.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum6.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

## | **Liquid Chromatography Mass Spectrometry (LC-MS)** {.notransition}

```{r out.width = "700px", echo = FALSE}
knitr::include_graphics("images/LCMS_Spectrum6.png")
```

<div style="position:absolute; left:50%; top:55%">
We gain an additional dimension:

- retention time.
</div>

<div style="position:absolute; left:50%; top:70%">
- LC-MS: analyze data along retention time.
</div>

## LC-MS data preprocessing

- Chromatographic peak detection
- Alignment
- Correspondence

## Chromatographic peak detection

- **Aim:** identify chromatographic peaks in the data.

```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/LCMS-data.png")
```

## Chromatographic peak detection

- **Aim:** identify chromatographic peaks in the data.

```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/LCMS-data-peaks.png")
```

## Chromatographic peak detection {.build}

<div>
- __centWave__ [Tautenhahn et al. BMC Bioinformatics, 2008]:
</div>
<div>
- Allows detection of peaks with different rt widths.

```{r out.width = "500px", echo = FALSE}
knitr::include_graphics("images/centWave-CWT.png")
```
</div>


## Chromatographic peak detection {.build}

- `MSnbase`: data import with `readMSData`.
- `xcms`: peak detection with `findChromPeaks` and algorithm-specific parameter
  object.

```{r echo = FALSE, warning = FALSE, message = FALSE}
library(xcms)
register(SerialParam(), default = TRUE) 
fls <- dir(system.file("sciex", package = "msdata"), full.names = TRUE)

## Define a data.frame with additional information on the files.
pd <- data.frame(file = basename(fls), injection_idx = c(1, 19),
                 sample = c("POOL_1", "POOL_2"), group = "POOL")
## Read the data
data <- readMSData(fls, pdata = new("NAnnotatedDataFrame", pd), 
                   mode = "onDisk")
data <- pickPeaks(smooth(data, method = "SavitzkyGolay", halfWindowSize = 6))
```

```{r message = FALSE, warning = FALSE}
cwp <- CentWaveParam(peakwidth = c(2, 10), snthresh = 5)
data <- findChromPeaks(data, param = cwp)
head(chromPeaks(data), n = 3)
```



## Alignment {.build}

<div>
- **Aim:** adjust differences in retention times between samples.
</div>
<div>
- Same analyte elutes at slightly different time between measurements.

```{r echo = FALSE, out.width = "370px"}
knitr::include_graphics("images/alignment.png")
```
</div>

<div>
- **Why?** Age of column, temperature ...
</div>

## Alignment {.build}

<div>
- Many algorithms available [Smith et al. Brief Bioinformatics 2013]
</div>

<div>
- `xcms`: `adjustRtime` function with `PeakGroupsParam` [Smith et al. Anal. chem.
    2006] or `ObiwarpParam` [Prince et al. Anal. chem. 2006].
</div>

<div>
```{r echo = FALSE, message = FALSE, warning = FALSE, fig.width = 8.5, fig.height = 3}
pdp <- PeakDensityParam(sampleGroups = data$group, bw = 1.8,
                        minFraction = 1, binSize = 0.02)
data <- groupChromPeaks(data, pdp)
pgp <- PeakGroupsParam(minFraction = 1, span = 0.6)
data <- adjustRtime(data, param = pgp)

par(mfrow = c(1, 2), mar = c(4, 4, 0.9, 0.2))
plot(chromatogram(data, aggregationFun = "sum", adjustedRtime = FALSE,
                  rt = c(70, 220)), main = "TIC, raw", peakType = "none")
plot(chromatogram(data, aggregationFun = "sum", rt = c(70, 220)),
     main = "TIC, adjusted", peakType = "none")
```
</div>

## Correspondence {.build}

- **Aim:** group peaks representing same ion species across samples.
- **Result:** matrix of abundances, rows *features*, columns samples.

```{r echo = FALSE, out.width = "700px"}
knitr::include_graphics("images/correspondence2_01.png")
```

## Correspondence

- **Aim:** group peaks representing same ion species across samples.
- **Result:** matrix of abundances, rows *features*, columns samples.

```{r echo = FALSE, out.width = "700px"}
knitr::include_graphics("images/correspondence2_02.png")
```

## Correspondence

- **Aim:** group peaks representing same ion species across samples.
- **Result:** matrix of abundances, rows *features*, columns samples.

```{r echo = FALSE, out.width = "700px"}
knitr::include_graphics("images/correspondence2_03.png")
```


## Correspondence {.build}

- `xcms`: `groupChromPeaks` with `NearestPeaksParam` [Katajamaa et
  al. Bioinformatics 2006] and `PeakDensityParam` [Smith et al. Anal. chem.
  2006].


## Correspondence

- `xcms`: `groupChromPeaks` with `NearestPeaksParam` [Katajamaa et
  al. Bioinformatics 2006] and `PeakDensityParam` [Smith et al. Anal. chem.
  2006].
- *Peak density* approach (for a given m/z slice):
- Identify regions along rt with high peak density, group peaks.

```{r echo = FALSE, out.width = "700px"}
knitr::include_graphics("images/correspondence2_density.png")
```

## Preprocessing result {.build}

- Numeric matrix with abundances.
- Normalization.
- Identification of features of interest.
- Annotation.

## Normalization {.build}

Account for:
<div>
- Sample-specific effects.
- Effects related to batch/measurement run.
- Injection order-dependent effects: specific to metabolite.
</div>

<div>
```{r echo = FALSE, out.width = "700px"}
knitr::include_graphics("images/FT07853.png")
```
</div>

## Normalization {.build}

<div>
- Good practice for experimental design:
  - QC samples measured repeatedly.
  - Internal standards.
  - Replicates.
  - Measurement of study samples in randomized order.
</div>
<div>
- Popular normalization methods:
  - RUV [De Livera et al. Anal. Chem. 2015] 
  - linear models [Wehrens et al. Metabolomics 2016]
  - linear and higher order models [Brunius et al. Metabolomics 2016].
</div>

## Annotation/Identification {.build}

- Feature != metabolite.

```{r echo = FALSE, warning = FALSE, message = FALSE}
data <- groupChromPeaks(data, param = PeakDensityParam(
                                  sampleGroups = data$group,
                                  bw = 1.8, binSize = 0.02, minFraction = 0.4))
res <- cbind(featureDefinitions(data), featureValues(data, value = "into"))
res <- res[, c(1, 4, 10, 11)]
colnames(res)[3:4] <- c("POOL_1", "POOL_2")
head(res, n = 4)
```

- Feature characterized by m/z and retention time.

## Annotation based on mass matching{.build}

<div>
- m/z is **not** the mass.
</div>
<div>
- Mass of an [M+H]+ ion: m/z - mass of 1 hydrogen.
</div>
<div>
- Different ions from the same compound: [M+H]+, [M+Na]+, ...
</div>
<div>
- Match mass against database.
  - The Human Metabolome Database (HMDB): <small>https://hmdb.ca</small>
  - Chemical Entities of Biological Interest: https://www.ebi.ac.uk/chebi
  - PubChem https://pubchem.ncbi.nlm.nih.gov
  - ...
- Will result in many hits.
</div>

## Improved Annotation {.build}

Annotate features based on m/z **and**:

<div>
- **retention time**: requires measurement of compound/standard on the same
  LC-MS setup.
</div>
<div>
- **MS2 spectrum**:
  - Requires LC-MS/MS data (DDA or DIA).
  - Reference spectrum has to be available in database.
</div>


## Afternoon metabolomics lab

- LC-MS data handling (`MSnbase`).
- LC-MS data preprocessing using `xcms`.

<div style="position:absolute; top:40%; left:35%;">
```{r, echo = FALSE, out.width = "240px"}
knitr::include_graphics("images/mass_spec_pkgs.png")
```
</div>
